File: progetto_algoritmi/LinAlg.cpp
#include "LinAlg.h"
namespace LinAlg {

// --- Funzioni matematiche ---
double to_radians(double degree) {
    return degree * (PI / 180.0f);
}

double sign(double value)
{ 
    return value > 0?1:-1; 
}

double to_degree(double angle) {
    return angle * (180.0f / PI);
}

double distance_between_vectors(double a, double b, double alpha) {
    return sqrt((a * a) + (b * b) - (2 * a * b * cos(to_radians(alpha))));
}

Vector2D distance_between_vectors(Vector2D a, Vector2D b) {
    return a - b;
}

Vector2D to_vector(double d, double alpha) {
    return Vector2D(d * cos(to_radians(alpha)), d * sin(to_radians(alpha)));
}

double opposite_angle(double adj_1, double adj_2, double opposite) {
    return acos(((adj_1 * adj_1) + (adj_2 * adj_2) - (opposite * opposite)) / (2 * adj_1 * adj_2));
}


double opposite_angle_vector(Vector2D adj_1, Vector2D adj_2, Vector2D opposite) {
    return acos(((adj_1.get_vnorm()) + (adj_2.get_vnorm()) - (opposite.get_vnorm())) / (2 * sqrt(adj_1.get_vnorm()) * sqrt(adj_2.get_vnorm()) ));
}



vector<double> get_all_angles(double a, double b, double c) {
    vector<double> r;
    r.push_back(opposite_angle(a, b, c));
    r.push_back(opposite_angle(b, c, a));
    r.push_back(opposite_angle(c, a, b));
    return r;
}

Vector2D get_avg_center(vector<Vector2D> vectors) {
    double avg_x = 0, avg_y = 0;
    int size = vectors.size();
    for (size_t i = 0; i < size; ++i) {
        avg_x += vectors[i].x;
        avg_y += vectors[i].y;
    }
    return Vector2D(avg_x / size, avg_y / size);
}

bool close_to(double a, double c, double e_tol) {
    return abs(a - c) < e_tol;
}

// --- Vector2D ---
Vector2D::Vector2D() : x(0), y(0), v_norm(0), v_degree(0) {}

Vector2D::Vector2D(double x, double y)
    : x(x), y(y), v_norm(evaluate_vnorm(x, y)), v_degree(to_degree(atan2(y, x))) {}

Vector2D::Vector2D(double radius, double alpha, int)
    : x(cos(to_radians(alpha)) * radius),
      y(sin(to_radians(alpha)) * radius),
      v_norm(radius), v_degree(alpha) {}

Vector2D& Vector2D::operator=(const Vector2D& v) {
    if (this != &v) {
        this->x = v.get_x();
        this->y = v.get_y();
        this->v_degree = v.v_degree;
        this->v_norm = v.v_norm;
    }
    return *this;
}

double Vector2D::get_vnorm() const { return v_norm; }
double Vector2D::get_vdegree() const {return v_degree < 0 ? 360+v_degree:v_degree; }
double Vector2D::get_x() const { return x; }
double Vector2D::get_y() const { return y; }

double Vector2D::evaluate_vnorm(double x, double y) {
    return sqrt(pow(x, 2) + pow(y, 2));
}

Vector2D Vector2D::operator-(const Vector2D& vector) const {
    return Vector2D(x - vector.x, y - vector.y);
}

Vector2D Vector2D::operator+(const Vector2D& vector) const {
    return Vector2D(x + vector.x, y + vector.y);
}

Vector2D Vector2D::operator*(const double& x) {
    return Vector2D(this->x * x, this->y * x);
}

void Vector2D::print_vector()
{
    Serial.println("PRINT VECTOR");

    Serial.print("X:");
    Serial.print(this->x);

    Serial.println("Y:");
    Serial.print(this->y);

    Serial.println("NORM:");
    Serial.print(this->v_norm);

    Serial.println("ANGLE:");
    Serial.print(this->v_degree);

    Serial.println(" END PRINT VECTOR");
}

LineParam::LineParam(const Vector2D& point_on_line, const Vector2D& direction)
{
    origin = point_on_line;
    dir    = make_unique_direction(direction);
}

LineParam::LineParam(const Vector2D& direction)
        : LineParam(Vector2D(0,0), direction){}

Vector2D LineParam::evaluate(double t) const { return origin + Vector2D(dir.x * t, dir.y * t);}

Vector2D LineParam::get_versor() const {return dir;}

double LineParam::slope() const {
    if (fabs(dir.x) < 1e-9)
        return 0; 
    return dir.y / dir.x;
}

Vector2D LineParam::make_unique_direction(const Vector2D& v) const
{
    double norm = v.get_vnorm();

    // caso direzione nulla → default verso (1,0)
    if (norm < 1e-12)
        return Vector2D(1, 0);

    // versore normale
    Vector2D u(v.x / norm, v.y / norm);

    // unificazione del verso
    // 1) se x < 0, inverti
    // 2) se x == 0 e y < 0, inverti
    bool x_zero = fabs(u.x) < 1e-9;

    if (u.x < 0 || (x_zero && u.y < 0)) {
        u.x = -u.x;
        u.y = -u.y;
    }

    return u;
}


} // namespace LinAlg

File: progetto_algoritmi/webot/devices/controllers/robot_controller/lin_alg.py
# lin_alg.py
import math
import numbers
PI = math.pi

def to_radians(degree):
    return degree * (PI / 180.0)

def to_degree(angle_rad):
    return angle_rad * (180.0 / PI)

def sign(value):
    return 1.0 if value > 0 else -1.0

def close_to(a, c, e_tol):
    return abs(a - c) < e_tol

# Vector2D equivalente al tuo C++
class Vector2D:
    def __init__(self, x=0.0, y=0.0, radius_alpha_flag=None):
        """
        - new Vector2D(x, y) -> cartesiane
        - new Vector2D(radius, alpha, 0) -> polari (alpha gradi)
        """
        if radius_alpha_flag is not None:
            # costruttore polar: (radius, alpha, int)
            radius = x
            alpha = y
            ang = to_radians(alpha)
            self.x = math.cos(ang) * radius
            self.y = math.sin(ang) * radius
            self.v_norm = radius
            self.v_degree = alpha
        else:
            self.x = x
            self.y = y
            self.v_norm = self.evaluate_vnorm(self.x, self.y)
            # atan2 -> radianti, poi gradi
            self.v_degree = to_degree(math.atan2(self.y, self.x))

    def evaluate_vnorm(self, x, y):
        return math.sqrt(x*x + y*y)

    def get_vnorm(self):
        return self.v_norm

    def get_vdegree(self):
        # normalizza come nel tuo C++
        deg = self.v_degree
        return deg if deg >= 0 else 360.0 + deg

    def get_x(self):
        return self.x

    def get_y(self):
        return self.y

    def __sub__(self, other):
        return Vector2D(self.x - other.x, self.y - other.y)

    def __add__(self, other):
        return Vector2D(self.x + other.x, self.y + other.y)

    def __mul__(self, k):
        if isinstance(k,numbers.Real):
            return Vector2D(self.x * k, self.y * k)
        elif isinstance(k,Vector2D):
            return self.get_x() * k.get_x() + self.get_y() * k.get_y() 
    
    

    def print_vector(self):
        print("PRINT VECTOR")
        print("X:", self.x)
        print("Y:", self.y)
        print("NORM:", self.v_norm)
        print("ANGLE:", self.v_degree)
        print("END PRINT VECTOR")

# distance_between_vectors (vector version)
def distance_between_vectors_vec(a: Vector2D, b: Vector2D):

    a.print_vector()
    b.print_vector
    
    return a - b

def multi_vector(v : Vector2D,u : Vector2D):
    return v.get_x() * u.get_x() 
    
def get_avg_center(vectors):
    if len(vectors) == 0:
        return Vector2D(0, 0)
    avg_x = sum(v.x for v in vectors) / len(vectors)
    avg_y = sum(v.y for v in vectors) / len(vectors)
    return Vector2D(avg_x, avg_y)

def get_all_angles(a, b, c):
    # calcola angoli opposti con legge dei coseni (restituisce in radianti)
    def opposite_angle(adj1, adj2, opp):
        return math.acos(((adj1*adj1) + (adj2*adj2) - (opp*opp)) / (2*adj1*adj2))
    return [opposite_angle(a, b, c), opposite_angle(b, c, a), opposite_angle(c, a, b)]


def dot_product(v : Vector2D,u: Vector2D):
    return  (v * u)  /  (v.get_vnorm() * u.get_vnorm()) 

def cross_product(v: Vector2D, u: Vector2D):
    return v.x * u.y - v.y * u.x

# LineParam come nel C++
class LineParam:
    def __init__(self, point_on_line=None, direction=None):
        if direction is None and point_on_line is not None:
            # costruttore LineParam(direction) -> origin = (0,0)
            self.origin = Vector2D(0, 0)
            self.dir = self.make_unique_direction(point_on_line)
        else:
            self.origin = point_on_line if point_on_line is not None else Vector2D(0, 0)
            self.dir = self.make_unique_direction(direction if direction is not None else Vector2D(1, 0))

    def evaluate(self, t):
        return self.origin + Vector2D(self.dir.x * t, self.dir.y * t)

    def get_versor(self):
        return self.dir

    def slope(self):
        if abs(self.dir.x) < 1e-9:
            return 0.0
        return self.dir.y / self.dir.x

    def make_unique_direction(self, v: Vector2D):
        norm = v.get_vnorm()
        if norm < 1e-12:
            return Vector2D(1, 0)
        u = Vector2D(v.x / norm, v.y / norm)
        x_zero = abs(u.x) < 1e-9
        if u.x < 0 or (x_zero and u.y < 0):
            u.x = -u.x
            u.y = -u.y
        return u
